### PR TITLE
Argobjs

### PR DESCRIPTION
**Pull request by Holado**
NOTE: This should not be merged until matching and complex object handling has been updated, at the very least.

Done:
* Object data in Verbs.conf replaced with enhanced and more detailed data 
* Argument handling updated, direct and indirect object kept separately, hopefully to facilitate matching.

TBD:
* Update matcher functions to look at all arguments to determine which verb frame it is
* Add handling for other error types, only OBJ-CASE for now
* Recognize when PPs are a part of a verb frame and when they denote time, place or other more abstract concepts
* Add handling for more complex objects, such as MWEs, fixed phrases, etc. For now, the handler only picks up the object case
* Merge subject information with verb frames to better separate different verb frames

### CODE REVIEW

File: src/reynir/Greynir.grammar [hunk 1]
```diff
@@ -4323,16 +4323,32 @@ NlStak_p3_ft_nf_hk → Manns_hk_nf
 NlStak_p3_ft_þf_hk → Manns_hk_þf
 NlStak_p3_ft_þgf_hk → Manns_hk_þgf
 
+NlStak_p3_et_nf_kk → VillaManns_nf
```
Comment (vthorsteinsson) [inline src/reynir/Greynir.grammar:4326]: The right place for these rules might be in the `#if(include_errors)` section at the end of Greynir.grammar.

File: src/reynir/verbframe.py [hunk 2]
```diff
@@ -107,14 +107,16 @@ def add_error(
...
         corrlist = corr.split(",")
         errlist = corrlist[0].split("-")
         errkind = errlist[0].strip()
         verb_with_cases = "_".join([verb] + args)
         if corrlist[0] == "OBJ-CASE":
+            VerbErrors.check_args(args)
             VerbErrors.OBJ_ERRORS[verb][args[0]] = corrlist[1].strip()
-        if errkind == "OBJ":
+        
+        """
```
Comment (vthorsteinsson) [inline src/reynir/verbframe.py:118]: Better to use a vertical column of # marks to comment out code (easy to add and delete in VSCode using multi-line selections, Ctrl+Alt+up/down)

File: src/reynir/verbframe.py [hunk 3]
```diff
@@ -210,15 +213,19 @@ class VerbFrame:
...
+        obj: str,
+        iobj: str,
         args: List[str],
         preps: Iterable[Tuple[str, str]],
         particle: Optional[str],
         score: Optional[int],
     ) -> None:
         self.verb = verb
-        assert 0 <= len(args) <= 2
+        #assert 0 <= len(args) <= 2
```
Comment (vthorsteinsson) [inline src/reynir/verbframe.py:224]: Move this assert to below the `self.cases` assignment below and do `assert 0 <= len(self.cases)<= 2`

File: src/reynir/verbframe.py [hunk 4]
```diff
@@ -248,86 +255,162 @@ def matches_pcl(self, particle: str) -> bool:
...
+        def get_particle(s: str) -> Tuple[str, Optional[str]]:
+            # Process particles, should only be one in each line
+            particle = None
+            ix = s.rfind("*")
+            if ix >= 0:
+                particle = s[ix:].strip()
+                s = s[0:ix].strip()
+                if " " in particle:
+                    raise ConfigError("Particle should only be one word")
+                elif len(particle) < 2:
```
Comment (vthorsteinsson) [inline src/reynir/verbframe.py:305]: `len(particle) < 1`?  (Or then simply `if not particle`)?

File: src/reynir/verbframe.py [hunk 5]
```diff
@@ -248,86 +255,162 @@ def matches_pcl(self, particle: str) -> bool:
...
+                # We expect something like 'af þgf', or possibly
+                # 'fyrir_hönd þf' (where the underscore needs to be replaced by a space)
+                # or 'milli ef_ft' (detailing the plural)
+                p = ap[ix].strip()
+                parg = p.split()
+                #if len(parg) != 2:
+                #    raise ConfigError("Preposition should have exactly one argument")
+                if parg[1] not in ALL_CASES and parg[1] not in SUBCLAUSES:
+                    parg[1] = REFLPRN.get(parg[1], parg[1])
+                    assert parg[1] is not None
```
Comment (vthorsteinsson) [inline src/reynir/verbframe.py:325]: This assert can be deleted, it never triggers

File: src/reynir/verbframe.py [hunk 6]
```diff
@@ -248,86 +255,162 @@ def matches_pcl(self, particle: str) -> bool:
...
+                    spl = parg[1].split("_")
+                    skip = ["gr", "ft", "est", "mst", "et", "kk", "kvk", "hk"]   # TODO Handle these variants
```
Comment (vthorsteinsson) [inline src/reynir/verbframe.py:327]: Convert to `frozenset()` and move to constant at beginning of file

File: src/reynir/verbframe.py [hunk 7]
```diff
@@ -248,86 +255,162 @@ def matches_pcl(self, particle: str) -> bool:
...
+                    while True:
```
Comment (vthorsteinsson) [inline src/reynir/verbframe.py:328]: Suggest:
`while spl and spl[-1].strip() in skip: spl = spl[:-1]`

File: src/reynir/verbframe.py [hunk 8]
```diff
@@ -248,86 +255,162 @@ def matches_pcl(self, particle: str) -> bool:
...
+                        if spl[-1].strip() in skip:
+                            spl = spl[:-1]
+                        else:
+                            break
+                    if spl[-1] not in ALL_CASES:
+                        raise ConfigError(
+                            "Preposition argument must have a case as its last variant"
```
Comment (vthorsteinsson) [inline src/reynir/verbframe.py:335]: As this works now, a preposition argument must have a case as its first variant, or its last variant before all the skipped variants.

File: src/reynir/verbframe.py [hunk 9]
```diff
@@ -248,86 +255,162 @@ def matches_pcl(self, particle: str) -> bool:
...
+                    parg[1] = REFLPRN.get(parg[1], parg[1])
+                    assert parg[1] is not None
+                    spl = parg[1].split("_")
+                    skip = ["gr", "ft", "est", "mst", "et", "kk", "kvk", "hk"]   # TODO Handle these variants
+                    while True:
+                        if spl[-1].strip() in skip:
+                            spl = spl[:-1]
+                        else:
+                            break
+                    if spl[-1] not in ALL_CASES:
```
Comment (vthorsteinsson) [inline src/reynir/verbframe.py:333]: `if not spl or spl[-1] not in ALL_CASES:` - spl may have become an empty list at this point.

File: src/reynir/verbframe.py (hunk 8 above)
Comment (vthorsteinsson) [reply src/reynir/verbframe.py:335]: ...and spl may have more than one remaining list item - what happens to the ones left before the last item?

File: src/reynir/verbframe.py [hunk 10]
```diff
@@ -248,86 +255,162 @@ def matches_pcl(self, particle: str) -> bool:
...
+                prepositions.append((parg[0].replace("_", " "), parg[1]))
+                ix += 1
+            return s, prepositions
+
+        def get_direct_object(s: str) -> Tuple[str, str]:
+            # Process direct object argument
+            op = s.split("|")
+            s = op[0]
+            obj: str = ""
+            if len(op) < 1 or len(op) > 2:
```
Comment (vthorsteinsson) [inline src/reynir/verbframe.py:347]: Slightly more concise: `if not 1 <= len(op) <= 2:`

File: src/reynir/verbframe.py [hunk 11]
```diff
@@ -248,86 +255,162 @@ def matches_pcl(self, particle: str) -> bool:
...
+                raise ConfigError("Verb should have zero or one direct object")
+            elif len(op) == 2:
```
Comment (vthorsteinsson) [inline src/reynir/verbframe.py:349]: It is slightly better to use if here, not elif - control has already flowed out of the statement if the exception was raised

File: src/reynir/verbframe.py [hunk 12]
```diff
@@ -248,86 +255,162 @@ def matches_pcl(self, particle: str) -> bool:
...
+                pobj = op[1].strip()
+                if pobj not in ALL_CASES and pobj not in SUBCLAUSES:
+                    pobj = pobj.split(" ")[0].strip()
+                    pobj = REFLPRN.get(pobj, pobj)
+                    assert pobj is not None
```
Comment (vthorsteinsson) [inline src/reynir/verbframe.py:354]: No need for this assert

File: src/reynir/verbframe.py [hunk 13]
```diff
@@ -248,86 +255,162 @@ def matches_pcl(self, particle: str) -> bool:
...
+                    spl = pobj.split("_")
+                    skip = ["gr", "ft", "est", "mst", "et", "kk", "kvk", "hk"]
```
Comment (vthorsteinsson) [inline src/reynir/verbframe.py:356]: Maybe this variant handling logic could be factored out into a separate function

File: src/reynir/verbframe.py [hunk 14]
```diff
@@ -248,86 +255,162 @@ def matches_pcl(self, particle: str) -> bool:
...
+                        )
+                    pobj = spl[-1]
+                obj = pobj
+            return s, obj
+
+        def get_indirect_object(s: str) -> Tuple[str, str]:
+            # Process indirect object argument
+            a = s.split()
+            if len(a) < 1:
+                raise ConfigError("Verb should have zero or one indirect objects")
```
Comment (vthorsteinsson) [inline src/reynir/verbframe.py:374]: Can a verb ever have more than one indirect object?

File: src/reynir/verbframe.py [hunk 15]
```diff
@@ -248,86 +255,162 @@ def matches_pcl(self, particle: str) -> bool:
...
+            #if len(a) < 1 or len(a) > 3:
+            #    raise ConfigError("Verb should have zero, one or two arguments")
+            verb = a[0]
+            iobj: str = ""
+            if not verb.isalpha():
+                raise ConfigError("Verb '{0}' is not a valid word".format(verb))
+            if len(a) > 1:
+                piobj = " ".join(a[1:]).strip()
```
Comment (vthorsteinsson) [inline src/reynir/verbframe.py:382]: The strip() is likely superfluous here.

File: src/reynir/verbframe.py [hunk 16]
```diff
@@ -248,86 +255,162 @@ def matches_pcl(self, particle: str) -> bool:
...
+                if piobj not in ALL_CASES and piobj not in SUBCLAUSES:
+                    piobj = piobj.split(" ")[0].strip()
```
Comment (vthorsteinsson) [inline src/reynir/verbframe.py:384]: Again, factor this into a separate function

File: src/reynir/verbframe.py [hunk 17]
```diff
@@ -248,86 +255,162 @@ def matches_pcl(self, particle: str) -> bool:
...
+                        raise ConfigError(
+                            "Indirect object must have a case as its last variant"
+                        )
+                    iobj = spl[-1]
+
+            return verb, iobj
+        s, score = get_score(s)
+        s, error = get_error(s)
+        s, particle = get_particle(s)
+        s, prepositions = get_preposition(s)
```
Comment (vthorsteinsson) [inline src/reynir/verbframe.py:404]: Nitpick: use consistent singular/plural

File: src/reynir/Greynir.grammar (hunk 1 above)
Comment (Holado) [reply src/reynir/Greynir.grammar:4326]: Done; added a description of each error rule chapter

File: src/reynir/verbframe.py (hunk 2 above)
Comment (Holado) [reply src/reynir/verbframe.py:118]: Fixed; unfortunately Ubuntu seems to override that keyboard shortcut :(

File: src/reynir/verbframe.py (hunk 3 above)
Comment (Holado) [reply src/reynir/verbframe.py:224]: Done

File: src/reynir/verbframe.py (hunk 5 above)
Comment (Holado) [reply src/reynir/verbframe.py:325]: Done

File: src/reynir/verbframe.py (hunk 6 above)
Comment (Holado) [reply src/reynir/verbframe.py:327]: Done

File: src/reynir/verbframe.py (hunk 7 above)
Comment (Holado) [reply src/reynir/verbframe.py:328]: Done

File: src/reynir/verbframe.py (hunk 9 above)
Comment (Holado) [reply src/reynir/verbframe.py:333]: Done

File: src/reynir/verbframe.py (hunk 10 above)
Comment (Holado) [reply src/reynir/verbframe.py:347]: Done

File: src/reynir/verbframe.py (hunk 11 above)
Comment (Holado) [reply src/reynir/verbframe.py:349]: Done

File: src/reynir/verbframe.py (hunk 12 above)
Comment (Holado) [reply src/reynir/verbframe.py:354]: Done

File: src/reynir/verbframe.py (hunk 14 above)
Comment (Holado) [reply src/reynir/verbframe.py:374]: A verb can have more than 2 objects (ég gaf honum hestinn að borða) but never more than one indirect object

File: src/reynir/verbframe.py (hunk 15 above)
Comment (Holado) [reply src/reynir/verbframe.py:382]: Done

File: src/reynir/verbframe.py (hunk 17 above)
Comment (Holado) [reply src/reynir/verbframe.py:404]: Noted, fixed

File: src/reynir/verbframe.py (hunk 4 above)
Comment (Holado) [reply src/reynir/verbframe.py:305]: Done

File: src/reynir/verbframe.py (hunk 13 above)
Comment (Holado) [reply src/reynir/verbframe.py:356]: Done, much clearer!

File: src/reynir/verbframe.py (hunk 16 above)
Comment (Holado) [reply src/reynir/verbframe.py:384]: Done

File: src/reynir/verbframe.py (hunk 8 above)
Comment (Holado) [reply src/reynir/verbframe.py:335]: For now, they are not handled. This will be fixed in the next iteration.

**Comment by Holado** [comment]
Review issues have been addressed. Pytest issues have yet to be resolved.

**Comment by Holado** [comment]
* Data has been refined further: nh and nhx polished, complex objects updated, comments resolved. Data has been annotated with comments on subjects for each frame (at least mostly), so the data is ready for merging the two frame structures.
* Only 'vanilla' verb frames are added at the moment. Adding PPs, particles, and more complex objects requires changing the grammar as a whole, as well as how we define verb terminals. PPs especially present a problem, as noted before.
* The matcher functions only look at the case for now, as only vanilla verb frames are present. Preparations have been made for more complex handling, and each case has been documented with comments for future handling.
* Tests were also run with ParsingTestPipe, and they give similar results as before, so the parsing capacity should not be too affected by the uprooting.

**Pull request merged**